quota/quota_test.go

```diff
@@ -0,0 +1,543 @@
+				},
+				"subgroup2": {
+					Name:        "subgroup2",
+					MemoryLimit: quantity.SizeMiB,
+					ParentGroup: "subgroup1",
+				},
+			},
+			comment: "deeply nested valid fully specified sub-groups",
+		},
+		{
+			grps: map[string]*quota.Group{
+				"foogroup": {
+					Name:        "foogroup",
+					MemoryLimit: quantity.SizeMiB,
+					SubGroups:   []string{"subgroup1"},
+				},
+				"subgroup1": {
+					Name:        "subgroup1",
+					MemoryLimit: quantity.SizeMiB,
+					ParentGroup: "foogroup",
+					SubGroups:   []string{"subgroup2"},
+				},
+				"subgroup2": {
+					Name:        "subgroup2",
+					MemoryLimit: quantity.SizeMiB,
+					// missing parent reference
+				},
+			},
+			err:     `group "subgroup2" does not reference necessary parent group "subgroup1"`,
+			comment: "deeply nested invalid fully specified sub-groups",
+		},
+		{
+			grps: map[string]*quota.Group{
+				"not-foogroup": {
+					Name:        "foogroup",
+					MemoryLimit: quantity.SizeMiB,
+				},
+			},
+			err:     `group has name "foogroup", but is referenced as "not-foogroup"`,
+			comment: "group misname",
+		},
+		{
+			grps: map[string]*quota.Group{
+				"foogroup": {
+					Name:        "foogroup",
+					MemoryLimit: quantity.SizeMiB,
+					SubGroups:   []string{"other-missing"},
+				},
+			},
+			err:     `missing group "other-missing" referenced as the sub-group of group "foogroup"`,
+			comment: "missing sub-group name",
+		},
+		{
+			grps: map[string]*quota.Group{
+				"foogroup": {
+					Name:        "foogroup",
+					MemoryLimit: quantity.SizeMiB,
+					ParentGroup: "other-missing",
+				},
+			},
+			err:     `missing group "other-missing" referenced as the parent of group "foogroup"`,
+			comment: "missing sub-group name",
+		},
+	}
+
+	for _, t := range tt {
+		comment := Commentf(t.comment)
+		err := quota.ResolveCrossReferences(t.grps)
+		if t.err != "" {
+			c.Assert(err, ErrorMatches, t.err, comment)
+		} else {
+			c.Assert(err, IsNil, comment)
+		}
+	}
+}
+
+func (ts *quotaTestSuite) TestAddAllNecessaryGroups(c *C) {
+	qs := &quota.QuotaGroupSet{}
+
+	// it should initially be empty
+	c.Assert(qs.AllQuotaGroups(), HasLen, 0)
+
+	grp1, err := quota.NewGroup("myroot", quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// add the group and make sure it is in the set
+	qs.AddAllNecessaryGroups(grp1)
+	c.Assert(qs.AllQuotaGroups(), DeepEquals, []*quota.Group{grp1})
+
+	// adding multiple times doesn't change the set
+	qs.AddAllNecessaryGroups(grp1)
+	qs.AddAllNecessaryGroups(grp1)
+	c.Assert(qs.AllQuotaGroups(), DeepEquals, []*quota.Group{grp1})
+
+	// add a new group and make sure it is in the set now
+	grp2, err := quota.NewGroup("myroot2", quantity.SizeGiB)
+	c.Assert(err, IsNil)
+	qs.AddAllNecessaryGroups(grp2)
+	c.Assert(qs.AllQuotaGroups(), DeepEquals, []*quota.Group{grp1, grp2})
+
+	// make a sub-group and re-add the root group - it will automatically add
+	// the sub-group without us needing to explicitly add the sub-group
```

Comment:
a bit unclear to me what usage pattern would use this in particular. In the end you need to add the group in some way? my preference is just to assume that we don't support updating subgroups while adding to QuotaGroupSet

Reply:
This is just a unit test we likely won't use it that way. You can see how we use this function in the bigger PR but essentially it's just that we might be modifying one group in the tree, and thus we want to ensure that all groups from the tree are ensured to be updated.

I agree that I don't know in what situation changing the root group would have an effect on changing the sub-groups but I don't see how it breaks anything to do this and it's simpler code